acnportal/acnsim/base.py

```diff
@@ -0,0 +1,473 @@
+                f"method. Serialized object may not load correctly. "
+                f"Write a to_dict method and re-dump, or write an "
+                f"appropriate from_dict method to accurately load.",
+                UserWarning
+            )
+            for key in unserialized_keys:
+                unserialized_attr = self.__dict__[key]
+                # Try calling the attr's to_registry method.
+                try:
+                    registry, context_dict = unserialized_attr.to_registry(
+                        context_dict=context_dict)
+                    attribute_dict[key] = registry['id']
+                    continue
+                except AttributeError:
+                    pass
+                # Try dumping the object using the native JSON
+                # serializer.
+                try:
+                    json.dumps(unserialized_attr)
+                except TypeError:
+                    warnings.warn(
+                        f"Attribute {key} could not be serialized. "
+                        f"Dumping the attribute's repr() "
+                        f"representation. This attribute will not be "
+                        f"fully loaded.",
+                        UserWarning
+                    )
+                    attribute_dict[key] = repr(unserialized_attr)
+                else:
+                    attribute_dict[key] = unserialized_attr
+
+        obj_dict = {'class': obj_type, 'attributes': attribute_dict}
+        context_dict[obj_id] = obj_dict
+
+        # Check versions of acnportal and certain dependencies.
+        acnportal_version = pkg_resources.require('acnportal')[0].version
+        dependency_versions = {'numpy': numpy.__version__,
+                               'pandas': pandas.__version__}
+
+        return ({'id': obj_id,
+                 'context_dict': context_dict,
+                 'version': acnportal_version,
+                 'dependency_versions': dependency_versions},
+                context_dict)
+
+    def to_dict(self, context_dict=None):
+        """ Converts the object's attributes into a JSON serializable
+        dict. Each ACN-Sim object defines this method differently.
+
+        Args:
+            context_dict (Dict[str, JSON Serializable]): Dict mapping
+                object ID's to object JSON serializable representations.
+
+        Returns:
+            Dict[str, JSON Serializable]: Dict of the object's
+                attributes, converted to JSON serializable forms.
+            context_dict (Dict[str, JSON Serializable]): Dict mapping
+                object ID's to object JSON serializable representations.
+                This is returned as context_dict may have been modified
+                by this method.
+        """
+        raise NotImplementedError
+
+    @classmethod
+    def from_json(cls, in_json):
+        """ Returns an ACN-Sim object loaded from in_registry. """
```

Comment:
should a file object be passed in as a string or file object?

Reply:
Pandas allows for either, depending on an input `path` argument. We can do it this way too.

Reply:
Added functionality to pass in a file path or file object to to/from_json.